targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/gpio_irq_api.c

```diff
@@ -0,0 +1,317 @@
+
+    // Do the same to IENB
+    adi_gpio_GetGroupInterruptPins(port, SYS_GPIO_INTB_IRQn, 1 << pin_number, &int_reg_val);
+
+    // clear the bit for the pin
+    int_reg_val &= ~(1 << pin_number);
+
+    // write the interrupt register
+    adi_gpio_SetGroupInterruptPins(port, SYS_GPIO_INTB_IRQn, int_reg_val);
+}
+
+
+/** Function to set the relevant interrupt enable bits in either the IENA and IENB registers
+ *  for the given GPIO pin.
+ */
+static void enable_pin_interrupt(ADI_GPIO_PORT port, uint32_t pin_number, IRQn_Type eIrq)
+{
+    uint16_t int_reg_val;
+
+    // Read the current interrupt enable register content
+    adi_gpio_GetGroupInterruptPins(port, eIrq, 1 << pin_number, &int_reg_val);
+
+    // set the bit for the pin
+    int_reg_val |= (1 << pin_number);
+
+    // write the interrupt register
+    adi_gpio_SetGroupInterruptPins(port, eIrq, int_reg_val);
+}
+
+
+/** Initialize the GPIO IRQ pin
+ *
+ * @param obj     The GPIO object to initialize
+ * @param pin     The GPIO pin name
+ * @param handler The handler to be attached to GPIO IRQ
+ * @param id      The object ID (id != 0, 0 is reserved)
+ * @return -1 if pin is NC, 0 otherwise
+ */
+int gpio_irq_init(gpio_irq_t *obj, PinName pin, gpio_irq_handler handler, uint32_t id)
+{
+    uint32_t port = pin >> GPIO_PORT_SHIFT;
+    uint32_t pin_num = pin & 0xFF;
+
+    // check for valid pin and ID
+    if ((pin == NC) || (id == 0))
+        return -1;
+
+    // make sure gpio driver has been initialized
+    if (!gpio_initialized) {
+        adi_gpio_Init(gpioMemory,ADI_GPIO_MEMORY_SIZE);
+        gpio_initialized = 1;
+    }
+
+    // save the handler
+    if (handler)
+        irq_handler = handler;
+
+    // disable the interrupt for the given pin
+    disable_pin_interrupt((ADI_GPIO_PORT)port, pin_num);
+
+    // set the port pin as input
+    adi_gpio_InputEnable(port, 1 << pin_num, true);
+
+    // save the ID for future reference
+    channel_ids[port][pin_num].id = id;
+    channel_ids[port][pin_num].event = IRQ_NONE;
+    channel_ids[port][pin_num].int_enable = 0;
+    obj->id = id;
+    obj->pinname = pin;
+
+    return 0;
+}
+
+/** Release the GPIO IRQ PIN
+ *
+ * @param obj The gpio object
+ */
+void gpio_irq_free(gpio_irq_t *obj)
+{
+    uint32_t port = obj->pinname >> GPIO_PORT_SHIFT;
+    uint32_t pin_num = obj->pinname & 0xFF;
+
+    // disable interrupt for the given pin
+    gpio_irq_disable(obj);
+
+    // clear the status table
+    channel_ids[port][pin_num].id = 0;
+    channel_ids[port][pin_num].event = IRQ_NONE;
+    channel_ids[port][pin_num].int_enable = 0;
+}
+
+/** Enable/disable pin IRQ event
+ *
+ * @param obj    The GPIO object
+ * @param event  The GPIO IRQ event
+ * @param enable The enable flag
+ */
+void gpio_irq_set(gpio_irq_t *obj, gpio_irq_event event, uint32_t enable)
+{
+    uint16_t int_polarity_reg;
+    uint32_t port = obj->pinname >> GPIO_PORT_SHIFT;
+    uint32_t pin_num = obj->pinname & 0xFF;
+
+    if (event == IRQ_NONE)
```

Comment:
`{` , also line 251 in this file, please align the code in HAL code files like this one